### ISSUE TITLE
RFC: Administration object are'nt much extensible.

### ISSUE THREAD

**Issue by mattiamanzati** [mobxjs/mobx-state-tree#28]
@mweststrate I have some concern about types and constructor functions (factories) in mobx-state-tree :/
The problem is that administration object are'nt much extensible right now.
There is no way right now to do something like "hey, take that factory administration object and enhance it like this.". 

For example I was wondering about on how to implement the "unionOf" introduced by #21 with the addition of a field in the snapshot "$$type" or similar which is the field type, and found no answer, since right now there is no way of taking the child administration object and enhance it.

So maybe I was wondering; since administration objects are striclty related to patch emitting/applying and knowing who owns that node, maybe we should move some of the operation to the type?

**Comment by mweststrate** [comment]
Ha yeah I see the problem, gonna think about it. Seems like we need to be able to express that factories are composed and hence multiple factories can be 'attached' to a single model instance

**Comment by mattiamanzati** [comment]
Yeah, or somehow the administration node provides the default snapshot/events that are then bubbled up to a factory method? I don't know exaclty how they should be handled. I would love to find a great separation of concerns that also allow some sort of internal flexibility in defining new types/administration objects.

**Comment by mweststrate** [comment]
Ok, I think I figured it out; we should keep Node, but not inherit from it. Instead put all the variations of the abstract methods in static `ITypeHandler`'s (e.g. `applySnapshot`). So the Node stores only things like path, parent and such. Besides that the node should store a reference to the *declared* factory (e.g. the union), and the *actual* factory (the concrete type). Factories should beside a config also keep a reference to these typehandlers (they are actually the utility methods of factories, but mixing stuff into functions is always a bit annoying). 

From there on we could do roughly things like: 

```
function applySnapshot(object, snapshot) {
  const node = getNode(object)
  node.declaredFactory.typeHandler.applySnaphot(node, object, snapshot) // can check / change the actual type
}

function getSnapshot(object) {
  const node = getNode(object)
  node.actualFactory.typeHandler.applySnaphot(node, object, snapshot) // can check / change the actual type
}
```

It can maybe be a bit more simple than that, but I think if we make sure that there is only one type of `Node`, this allows for easier reuse of instances, and from the node all type specific logic could be delegated to either the declared or actual type. I think that makes it a bit cleaner. Probably the configuration of a factory then doesn't need to be stored as separate object anymore, but we could just associate a new instance of a certain typeHandler with a factory method.

I think I can do that refactoring this week, unless you beat me to eat.

**Comment by mattiamanzati** [comment]
Nice! This is almost the same thing I was thinking of, so this should be a reasonable path! Only few questions I was reasoning about and prevented me from posting this solution:

- Why do you restore path from actual factory and apply from declared one?
- Should'nt all the methods invoke the declared factory?
- Notice cases like this:

const Cube = createFactory({
   width: 0
})

const Sphere = createFactory({
   radius: 0
})

const Union = unionOf(Cube, Sphere)
const Tree = createFactory({
   shape: Union
})

// notice I declare as Cube
const cube = Cube()

// I cast to Union
const doc = Tree()
doc.shape = cube

// Should be Union.getSnapshot or Cube.getSnapshot?
getSnapshot(doc) // maybe this will use the Union from the tree
getSnapshot(cube) // and this the cube

PS: You can use Type.dispatch(snapshotOrNode) to retrive the actual type from the union type for example, so maybe storing the actual type ;)

**Comment by mweststrate** [comment]
I think it is important that things like `clone` use the declared type. If you just take the actual type and clone with that, you would after that not able to update the snapshot of the clone that matches the declared type but not the actual one. For example if you copy an object that has `x: maybe(String)` and it's current value is null, the `x` cannot be updated with a string in the clone, if the actual type instead of the declared one is used.

Yeah probably all methods should use the declared factory indeed.

Interesting example, I think the first should indeed create a Tree with Union, and the second example indeed use `Cube`, because that is the declared and actual type of `cube`. So maybe a node should store only its 'actual' type, and not the declared type at all, but we should just make sure that when a snapshot is applied to a collection, the declared type should be checked. 

That is straightforward to implement, although it has some tricky edge cases:

```
applySnapshot(cube, { radius: 3 }) // error, expected cube
applySnapshot(doc.shape, { radius: 3 }) // error, expected cube
applySnapshot(doc, { shape: { radius: 3 } }) // ok
doc.shape = { radius: 3 } // ok, replaces the cube, creates sphere
```

I think that case 1 throwing is correct, and therefor 2 for consistency as well. If you run into such an edge case, you can always apply a (partial) snapshot to the parent

**Comment by mattiamanzati** [comment]
Seems totally fine and an expected behaviour! I think that if we mention it in the docs we will be fine! :)

**Comment by mweststrate** [comment]
created WIP branch `feature/21-type-system`

**Comment by mattiamanzati** [comment]
Solved by #32
